packages/components/nodes/chains/GraphCypherQAChain/GraphCypherQAChain.ts

```diff
@@ -0,0 +1,273 @@
+
+        if (qaPrompt instanceof PromptTemplate) {
+            qaPromptTemplate = new PromptTemplate({
+                template: qaPrompt.template as string,
+                inputVariables: qaPrompt.inputVariables
+            })
+        }
+
+        if (!cypherModel || !qaModel) {
+            if (!model) {
+                throw new Error('Model or is required when Cypher Model or QA Model are not provided')
+            }
+        }
+
+        // validate cypherPromptTemplate is PromptTemplate and qaPromptTemplate must be exists
+        if (cypherPromptTemplate instanceof PromptTemplate && !qaPromptTemplate) {
+            throw new Error('QA Prompt is required when Cypher Prompt is a Prompt Template')
+        }
+
+        // Validate required variables in prompts
+        if (!cypherPromptTemplate.inputVariables.includes('schema') || !cypherPromptTemplate.inputVariables.includes('question')) {
+            throw new Error('Cypher Generation Prompt must include {schema} and {question} variables')
+        }
+
+        const chain = GraphCypherQAChain.fromLLM({
+            llm: model,
+            qaLLM: qaModel,
+            cypherLLM: cypherModel,
+            graph,
+            cypherPrompt: cypherPromptTemplate,
+            qaPrompt: omitQaPrompt ? undefined : qaPromptTemplate,
+            returnDirect,
+            returnIntermediateSteps
+        })
+
+        return chain
+    }
+
+    async run(nodeData: INodeData, input: string, options: ICommonObject): Promise<string | object> {
+        const chain = nodeData.instance as GraphCypherQAChain
+        const memory = nodeData.inputs?.memory as FlowiseMemory
+        const moderations = nodeData.inputs?.inputModeration as Moderation[]
+
+        const shouldStreamResponse = options.shouldStreamResponse
+        const sseStreamer: IServerSideEventStreamer = options.sseStreamer as IServerSideEventStreamer
+        const chatId = options.chatId
+
+        // Handle input moderation if configured
+        if (moderations && moderations.length > 0) {
+            try {
+                input = await checkInputs(moderations, input)
+            } catch (e) {
+                await new Promise((resolve) => setTimeout(resolve, 500))
+                if (shouldStreamResponse) {
+                    streamResponse(sseStreamer, chatId, e.message)
+                }
+                return formatResponse(e.message)
+            }
+        }
+
+        const obj = {
+            query: input
+        }
+
+        const loggerHandler = new ConsoleCallbackHandler(options.logger)
+        const callbackHandlers = await additionalCallbacks(nodeData, options)
+        let callbacks = [loggerHandler, ...callbackHandlers]
+
+        if (process.env.DEBUG === 'true') {
+            callbacks.push(new LCConsoleCallbackHandler())
+        }
+
+        try {
+            let response
+            if (shouldStreamResponse) {
+                const handler = new CustomChainHandler(sseStreamer, chatId)
```

Comment:
We use `CustomChainHandler` to stream the LLM output back to UI. However, this chain has 2 chains under the hood, this will then stream both LLM outputs back to UI, but we are only interested in the latter one.

To prevent this from happening, `CustomChainHandler` takes in a parameter called `skipK`. This will skip streaming for first K numbers of handleLLMStart event. In this scenario, you can specify 2 for that. And you will be able to see only the last LLM response will be streamed back.

By setting skipK to 2, it works without `returnDirect`. But when `returnDirect` is turned on, only Cypher Chain LLM is executed, not the QA Chain. Since we set it to 2, the LLM output from Cypher Chain will be skipped, causing nothing to stream back to UI.

To solve this, we can just directly get the final answer `let result = response?.result`, and stream the whole response back `if (result) streamResponse(sseStreamer, chatId, result)`

Reply:
thanks for the explication, the response has been fixed